src/components/ExportDiscourseContext.tsx

```diff
@@ -0,0 +1,393 @@
+import React, { useState } from "react";
+import ReactDOM from "react-dom";
+import {
+  Button,
+  Classes,
+  Dialog,
+  Label,
+  Slider,
+  Tab,
+  Tabs,
+  Tooltip,
+  Callout,
+  Collapse,
+} from "@blueprintjs/core";
+import renderOverlay, {
+  RoamOverlayProps,
+} from "roamjs-components/util/renderOverlay";
+import getPageUidByPageTitle from "roamjs-components/queries/getPageUidByPageTitle";
+import getPageTitleByPageUid from "roamjs-components/queries/getPageTitleByPageUid";
+import getPageTitlesandBlockUidsReferencingPage from "roamjs-components/queries/getPageTitlesAndBlockUidsReferencingPage";
+import getLinkedPageTitlesUnderUid from "roamjs-components/queries/getLinkedPageTitlesUnderUid";
+import isDiscourseNode from "../utils/isDiscourseNode";
+import getDiscourseContextResults from "../utils/getDiscourseContextResults";
+import { render as exportRender } from "./Export";
+import { Result } from "../utils/types";
+
+const getInboundReferences = async (uid: string) => {
+  // TODO: Optimize this with custom query to take in and return Result[]
+  const pageTitle = getPageTitleByPageUid(uid);
+  const refs = getPageTitlesandBlockUidsReferencingPage(pageTitle);
+  const refPageUids = refs.map((r) => getPageUidByPageTitle(r.title));
+  return refPageUids;
+};
+
+const getOutboundReferences = async (uid: string) => {
+  // TODO: Optimize this with custom query take in and return Result[]
+  const refPageTitles = getLinkedPageTitlesUnderUid(uid);
+  const refPageUids = refPageTitles.map((title) =>
+    getPageUidByPageTitle(title)
+  );
+  return refPageUids;
+};
+
+const getReferencesByDegree = async (
+  initialUids: string[],
+  degrees: number,
+  getReferences: (uid: string) => Promise<string[]>
+): Promise<Set<string>> => {
+  let currentUids = [...initialUids];
+  const visitedUids = new Set<string>();
+
+  for (let i = 0; i < degrees; i++) {
```

Comment:
See if we could avoid this loop

`[:find [pull ?b [:block/uid :block/string {:block/refs 3}]] :where [?b :block/uid "dq4_gLsOG"]]`

Reply:
Not super happy with where I got so far, so I'm going to defer this to later.
Also, I don't see a way to query all of the inbound references without a loop.

Here's outbound so far
``` typescript
const getOutBoundReferencesQuery = async (uid: string, depth: number) => {
  //TODO run this for each initial uid
  const adjustedDept = depth + 1
  const query = window.roamAlphaAPI.q(`
    [:find (pull ?parent 
            [:block/uid 
            :node/title
            {:block/children ${adjustedDept} 
              :block/refs ${adjustedDept}}]) 
    :where [?parent :block/uid "${uid}"]]
  `)[0][0] as []

  const extractUids = (node: any[]) => {
    let uids: string[] = [];

    function traverse(node: any) {
      if (Array.isArray(node)) {
        node.forEach(innerNode => traverse(innerNode));
      } else if (typeof node === 'object' && node !== null) {
        if ('uid' in node && 'title' in node) {
          uids.push(node.uid);
        }

        Object.keys(node).forEach(key => {
          if (typeof node[key] === 'object') {
            traverse(node[key]);
          }
        });
      }
    }

    traverse(node);
    return uids;
  }
  const uids = extractUids(query)
  return uids
}
```